### ISSUE TITLE
[BUG] Marker vanishes.

### ISSUE THREAD

**Issue by BillPitcher** [ironmonk88/monks-combat-marker#7]
The marker sometimes vanishes, when changing scene and doesn't return until a page refresh.

**To Reproduce**
Very difficult to Reproduce

**Please complete as much of the following information as possible:**
 - Browser Chrome and Standalone App
 - Foundry Version 10.291
 - System D&D5e
 - Module version Current
 - Are there any errors in the console, no errors seen

**Comment by nschoenwald** [comment]
I have the same issue sporadically, but it is hard to reproduce, like you say.

**Comment by Sisimshow** [comment]
My group and I are basically having this issue constantly. It seems to happen most often when we transition scenes. We use the Stairway module a lot, it's possible transitioning scenes with a Stairway makes it likely to happen for us. Feel free to message me if you need me to replicate the issue, it's driving us mad lol. 

That said, the module is awesome :)

**Comment by STBaf** [comment]
I can add here: Running on latest v11.307 and have Combat Marker + Stairways active. Can reproduce that we loose the marker functionality after the first combat (first combat = marker is there, second or more combats: no more marker). Reloading fixes this for one combat, after that same problem.
Have disabled the Stairways module, seem to have fixed the behavior, after this we had combat markers in second and further combats (but we did only some tests yet, not a lot of tests).
Seem Stairways is in some way interfering with the combat marker.

**Comment by git-kor-av** [comment]
I have the same issue sporadically as well and have not found a way to reproduce it consistently.  It does seem to have something to do with changing scenes.  There are no errors in the console.log when this happens.

My environment seems quite different from other players though.  I have Foundry 11.308 with PF2 5.3.2.  I do not have the Stairways module.

**Comment by BillPitcher** [comment]
An "interesting" thing I've noticed is when I refresh the page to get the marker to show again, it requires me "Gamemaster" to login again. Maybe there is a cache being timed out. Will continue to monitor, as I've just done all the latest updates.

**Comment by BillPitcher** [comment]
I'm thinking when returning to a scene with a running combat the toggleTurnMarker(token, visible) is being call before the canvas is drawn and because the token is not visible it removes the marker.
It seems to be a timing problem when sometimes it's fine and toggleTurnMarker is called late but others early and it just stops working. Very frustrating as it's become a key plugin for my games. Hope this helps.

**Comment by ironmonk108** [comment]
I'm really unable to replicate this.  I tried switching scenes and having Stairways enabled, and the marker doesn't disappear.

Please bear in mind that if you move a token between scenes, it's not the original token, and the marker won't be added to it automatically.

But other than that, I'm at a loss.

**Comment by BillPitcher** [comment]
Thanks for looking into this, I've spent much of today chasing this down. I have to say you're a wizard, your code is way above my pay-grade.

What I've got down to is that for some reason one of the ldmarkers in the markerCache has been marked as destroyed, it's key: the texture file name and the object are still there so the animateMarkers loop still fires but the ldmarker doesn't do anything because it's in the destroyed = true state.

The question still remains why, my current thoughts are that when the canvasReady fires it's not seeing the/a Tokens and destroying the ldmarker. The problem is a timing problem, sometimes Foundry has to sit at a different scene for quite some time before coming back to the ongoing combat breaks.

Interestingly once it's broken in 1 scene it's broken for them all, but I suspect that's because it's the users token that breaks and it's using the cached ldmarker. 

Any guidance to get to the bottom of this would be greatly appreciated.

``` JS
modules/monks-combat-marker/markers/marker02.webp : 
B
baseTexture : 
qe
alphaMode : 1
anisotropicLevel : 0
cacheId : null
destroyed : true
dirtyId : 0
dirtyStyleId : 0
```

**Comment by ironmonk108** [comment]
Huh.  Well... that's interesting.
This is going to be a really hard one to fix, not being able to replicate it... but 11.05 is out with some changes, and I'm crossing my fingers that you've managed to narrow it down.

**Comment by BillPitcher** [comment]
Yep, something is cleaning the PIXI.Sprite up, and I'm thinking its not monks-combat-marker.
I commented out every call to destroy() and left it for 20mins when returning to the scene it had been marked as destroyed.

I guess the next step in a work around that will check if it has been destroyed and if it has been rebuild it.

As you may have guessed I use this all the time. I often have many creatures with the same name in the Tracker and having the hi-lighted  as we work through combat makes combat proceed smoothly. It is the best, must have module!

**Comment by ironmonk108** [comment]
It could be a long shot... but you might be able to track it down running FTC?
And could be a reason I'm not able to replicate it, if I don't have a certain module installed.
Did 11.05 help at all?

**Comment by BillPitcher** [comment]
I can now predictably and repeatably cause this problem.
On a scene start a combat and note that the marker is showing.
Move to a scene with No combat
Wait......
Return to any scene with a combat and the markers are gone.

What I think is happening.
When a Sprite gets made it uploads it's Texture to the graphics card. This texture is a single item and is used by all sprites that use that texture. When the Sprite/Texture is not being used, as in when on a scene with no combat markers after a while it is destroyed. I think the "Wait" is dependant on the need for GPU ram. I use both maps and overlays and have a lot of hidden tokens, walls,  lights and tiles, I guess if any of these use GPU ram then it would speed up the cleanup process. I find that Chrome does this quicker than the standalone App again pointing to ram constraints.
Once the "unused" textures have been destroyed, the Sprites continue to "work" but they show nothing, and so there is no errors.

What I'm currently doing is in Hooks.on("canvasReady", function (canvas) checking MonksCombatMarker.markerCache for any Sprites with destroyed Textures, if it finds one it resets the whole markerCache = {}. This forces recreation of the Sprite and texture and all is good.

I like this solution as it should have no unwanted side effects and does nothing if not needed. 2 lines of code for every day spent getting a workaround.

```JS
	// Fix for Pixi Textures being Destroyed (Bill)
    for (var prop in MonksCombatMarker.markerCache) {
    	let holdSprite =  MonksCombatMarker.markerCache[prop];
    	if (holdSprite.baseTexture.destroyed) {
    		MonksCombatMarker.markerCache = {};
    		break;
    	}
    }
```

**Comment by BillPitcher** [comment]
Just a quick update, I've found this fix worked and i haven't had any problems. I can confirm Pixi does do a Texture Garbage Collection  of unused textures according to there documentation so I don't think Foundry can fix this without turning off the GC. Digging a bit deeper I need to destroy the Sprites also. So I've refactored my fix:

in the MonksCombatMarker class i would add these 2 functions
```JS
    static markerCacheNeedsDestroying() {
    	for (var prop in MonksCombatMarker.markerCache) {
    		let holdSprite =  MonksCombatMarker.markerCache[prop];
    		if (holdSprite.baseTexture.destroyed) {
    			return true;
    		}
    	}
    	return false
    }
    
	static markerCacheDestroy() {
		for (var prop in MonksCombatMarker.markerCache) {
			let holdSprite =  MonksCombatMarker.markerCache[prop];
			holdSprite.destroy()
		}
		MonksCombatMarker.markerCache = {}
    }
```

And in the Hooks.on("canvasReady", function (canvas) 
```JS
	// Fix for Pixi Textures being Destroyed 
	if (MonksCombatMarker.markerCacheNeedsDestroying() ) {
		MonksCombatMarker.markerCacheDestroy();
	}
	
```

YMMV
have a great day
Bill
